bentoml/utils/tensorflow.py

```diff
@@ -0,0 +1,214 @@
+    except ImportError:
+        raise MissingDependencyException(
+            "Tensorflow package is required to use TfSavedModelArtifact"
+        )
+
+    return m.signatures.get(tf.compat.v2.saved_model.DEFAULT_SERVING_SIGNATURE_DEF_KEY)
+
+
+def cast_tensor_by_spec(_input, spec):
+    '''
+    transform dtype & shape following spec
+    '''
+    try:
+        import tensorflow as tf
+    except ImportError:
+        raise MissingDependencyException(
+            "Tensorflow package is required to use TfSavedModelArtifact"
+        )
+
+    if not _isinstance(spec, "TensorSpec"):
+        return _input
+
+    if _isinstance(_input, ["Tensor", "EagerTensor"]):
+        # TensorFlow issue #43038
+        # pylint: disable=unexpected-keyword-arg, no-value-for-parameter
+        return tf.cast(_input, dtype=spec.dtype, name=spec.name)
+    else:
+        return tf.constant(_input, dtype=spec.dtype, name=spec.name)
+
+
+def _pretty_format_function_call(base, name, arg_names):
+    if arg_names:
+        part_sigs = ', '.join(f"{k}" for k in arg_names)
+    else:
+        part_sigs = ''
+
+    if name == "__call__":
+        return f"{base}({part_sigs})"
+    return f"{base}.{name}({part_sigs})"
+
+
+def _pretty_format_positional(positional):
+    return "Positional arguments ({} total):\n    * {}".format(
+        len(positional), "\n    * ".join(str(a) for a in positional)
+    )
+
+
+def pretty_format_function(function, obj="<object>", name="<function>"):
+    ret = ''
+    outs = get_output_signature(function)
+    sigs = get_input_signatures(function)
+    arg_names = get_arg_names(function)
+
+    if hasattr(function, 'function_spec'):
+        arg_names = function.function_spec.arg_names
+    else:
+        arg_names = function._arg_keywords
+
+    ret += _pretty_format_function_call(obj, name, arg_names)
+    ret += '\n------------\n'
+
+    signature_descriptions = []
+
+    for index, sig in enumerate(sigs):
+        positional, keyword = sig
+        signature_descriptions.append(
+            "Arguments Option {}:\n  {}\n  Keyword arguments:\n    {}".format(
+                index + 1, _pretty_format_positional(positional), keyword
+            )
+        )
+
+    ret += '\n\n'.join(signature_descriptions)
+    ret += f"\n\nReturn:\n  {outs}\n\n"
+    return ret
+
+
+def pretty_format_restored_model(model):
+    part_functions = ""
+
+    restored_functions = get_restored_functions(model)
+    for name, func in restored_functions.items():
+        part_functions += pretty_format_function(func, "model", name)
+        part_functions += "\n"
+
+    serving_default = get_serving_default_function(model)
+    if serving_default:
+        part_functions += pretty_format_function(
+            serving_default, "model", "signature['serving_default']"
+        )
+        part_functions += "\n"
+
+    if not restored_functions and not serving_default:
+        return (
+            "Found no avaliable functions. You can use `tf.function` "
```

Comment:
"No serving function was found in the saved model. In the model implementation, use `tf.function` decorator to mark the method needed for model serving. Find more details in related TensorFlow docs here https://www.tensorflow.org/api_docs/python/tf/saved_model/save"